python/src/scipp/plot/plot_2d.py

```diff
@@ -256,111 +277,222 @@ def update_axes(self):
-                vslice.masks["all"] = sc.Variable(dims=mslice_dims,
-                                                  values=mslice.values)
+            data=sc.Variable(dims=[
+                self.xyrebin[self.dim_to_xy[self.vslice.dims[0]]].dims[0],
+                self.xyrebin[self.dim_to_xy[self.vslice.dims[1]]].dims[0]
+            ],
+                             values=self.vslice.values,
+                             unit=sc.units.counts))
+
+        # Also include the masks
+        if self.params["masks"][self.name]["show"]:
+            mslice_dims = []
+            for dim in self.mslice.dims:
+                if dim == self.button_dims[0]:
+                    mslice_dims.append(self.xyrebin["y"].dims[0])
+                elif dim == self.button_dims[1]:
+                    mslice_dims.append(self.xyrebin["x"].dims[0])
+                else:
+                    mslice_dims.append(dim)
+
+            dslice.masks["all"] = sc.Variable(dims=mslice_dims,
+                                              values=self.mslice.values)
 
         # Scale by bin width and then rebin in both directions
-        if is_not_linspace["x"]:
-            vslice *= self.xywidth["x"]
-            vslice = sc.rebin(vslice, self.xyrebin["x"].dims[0],
-                              self.xyrebin["x"])
-        if is_not_linspace["y"]:
-            vslice *= self.xywidth["y"]
-            vslice = sc.rebin(vslice, self.xyrebin["y"].dims[0],
-                              self.xyrebin["y"])
-
-        # Remember to replace masks slice after rebin
-        if np.any(list(is_not_linspace.values())):
-            if self.params["masks"][self.name]["show"]:
-                mslice = vslice.masks["all"]
+        dslice *= self.xywidth["x"] * self.xywidth["y"]
+        # The order of the dimensions that are rebinned matters if 2D coords
+        # are present. We must rebin the base dimension of the 2D coord first.
+        xy = "yx"
+        if len(dslice.coords[self.button_dims[1]].dims) > 1:
+            xy = "xy"
+        dslice = sc.rebin(dslice, self.xyrebin[xy[0]].dims[0],
+                          self.xyrebin[xy[0]])
+        dslice = sc.rebin(dslice, self.xyrebin[xy[1]].dims[0],
+                          self.xyrebin[xy[1]])
+
+        # Use Scipp's automatic transpose to match the image x/y axes
+        arr = sc.DataArray(coords={
+            self.xyrebin["x"].dims[0]: self.xyrebin["x"],
+            self.xyrebin["y"].dims[0]: self.xyrebin["y"],
+        },
+                           data=sc.Variable(dims=self.button_dims,
+                                            values=np.ones([
+                                                self.xyrebin["y"].shape[0] - 1,
+                                                self.xyrebin["x"].shape[0] - 1
+                                            ]),
+                                            dtype=dslice.dtype,
+                                            unit=sc.units.one))
+        arr *= dslice
```

Comment:
Easier to use https://scipp.github.io/generated/scipp.transpose.html#scipp.transpose by hand?

Reply:
yes! much easier.
This is a recent addition, right?

Reply:
Has existed for a long time, but Python bindings where missing, Owen added them earlier this week.